homeassistant/components/osramlightify/light.py

```diff
@@ -4,270 +4,410 @@
-            red, green, blue = \
-                color_util.color_hs_to_RGB(*kwargs[ATTR_HS_COLOR])
-            self._luminary.set_rgb(red, green, blue, transition)
+            self._hs_color = kwargs[ATTR_HS_COLOR]
+            self._rgb_color = color_util.color_hs_to_RGB(*self._hs_color)
+            self._xy_color = color_util.color_hs_to_xy(*self._hs_color)
+            self._luminary.set_rgb(*self._rgb_color, transition)
+        elif ATTR_RGB_COLOR in kwargs:
+            self._rgb_color = kwargs[ATTR_RGB_COLOR]
+            self._hs_color = color_util.color_RGB_to_hs(*self._rgb_color)
+            self._xy_color = color_util.color_RGB_to_xy(*self._rgb_color)
+            self._luminary.set_rgb(*self._rgb_color, transition)
+        elif ATTR_XY_COLOR in kwargs:
+            self._xy_color = kwargs[ATTR_XY_COLOR]
+            self._rgb_color = color_util.color_xy_to_RGB(*self._xy_color)
+            self._xy_color = color_util.color_xy_to_hs(*self._xy_color)
+            self._luminary.set_rgb(*self._rgb_color, transition)
+        elif ATTR_COLOR_NAME in kwargs:
+            self._rgb_color = color_util.color_name_to_rgb(
+                kwargs[ATTR_COLOR_NAME])
+            self._hs_color = color_util.color_RGB_to_hs(*self._rgb_color)
+            self._xy_color = color_util.color_RGB_to_xy(*self._rgb_color)
+            self._luminary.set_rgb(*self._rgb_color, transition)
 
         if ATTR_COLOR_TEMP in kwargs:
-            color_t = kwargs[ATTR_COLOR_TEMP]
-            kelvin = int(color_temperature_mired_to_kelvin(color_t))
-            self._luminary.set_temperature(kelvin, transition)
-
-        if ATTR_EFFECT in kwargs:
-            effect = kwargs.get(ATTR_EFFECT)
-            if effect == EFFECT_RANDOM:
-                self._luminary.set_rgb(
-                    random.randrange(0, 255), random.randrange(0, 255),
-                    random.randrange(0, 255), transition)
+            self._color_temp = kwargs[ATTR_COLOR_TEMP]
+            self._luminary.set_temperature(
+                int(color_util.color_temperature_mired_to_kelvin(
+                    self._color_temp)), transition)
+        elif ATTR_KELVIN in kwargs:
+            self._color_temp = color_util.color_temperature_kelvin_to_mired(
+                kwargs[ATTR_KELVIN])
+            self._luminary.set_temperature(kwargs[ATTR_KELVIN], transition)
+
+        self._is_on = True
+        if ATTR_BRIGHTNESS in kwargs:
+            self._brightness = kwargs[ATTR_BRIGHTNESS]
+            self._luminary.set_luminance(int(self._brightness / 2.55),
+                                         transition)
+        elif ATTR_BRIGHTNESS_PCT in kwargs:
+            self._brightness = int(kwargs[ATTR_BRIGHTNESS_PCT] * 2.55)
+            self._luminary.set_luminance(int(kwargs[ATTR_BRIGHTNESS_PCT]),
+                                         transition)
+        else:
+            self._luminary.set_onoff(True)
 
         self.schedule_update_ha_state()
```

Comment:
Also remove `schedule_update_ha_state()`, that happens automatically after `turn_on()`.

Reply:
Sure.